TulparAudio/internal/source/SourceCollection.cpp

```diff
@@ -0,0 +1,1097 @@
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set relative: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+bool SourceCollection::IsSourceLooping(SourceHandle source) const
+{
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALint result = AL_FALSE;
+
+    alGetSourcei(static_cast<ALuint>(source), AL_LOOPING, &result);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        result = AL_FALSE;
+        LOG_AUDIO->Warning("Source #{}: get looping: {:#x}", source, alErr);
+    }
+
+    return result == AL_TRUE;
+}
+
+bool SourceCollection::SetSourceLooping(SourceHandle source, bool flag)
+{
+    LOG_AUDIO->Debug("Source #{}: set looping {}", source, flag);
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcei(static_cast<ALuint>(source), AL_LOOPING, (flag ? AL_TRUE : AL_FALSE));
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set looping: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+float SourceCollection::GetSourcePitch(SourceHandle source) const
+{
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALfloat result = 0.0f;
+
+    alGetSourcef(static_cast<ALuint>(source), AL_PITCH, &result);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        result = -1.0f;
+        LOG_AUDIO->Warning("Source #{}: get pitch: {:#x}", source, alErr);
+    }
+
+    return result;
+}
+
+bool SourceCollection::SetSourcePitch(SourceHandle source, float value)
+{
+    LOG_AUDIO->Debug("Source #{}: set pitch {}", source, value);
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcef(static_cast<ALuint>(source), AL_PITCH, value);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set pitch: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+float SourceCollection::GetSourceGain(SourceHandle source) const
+{
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALfloat result = 0.0f;
+
+    alGetSourcef(static_cast<ALuint>(source), AL_GAIN, &result);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        result = -1.0f;
+        LOG_AUDIO->Warning("Source #{}: get gain: {:#x}", source, alErr);
+    }
+
+    return result;
+}
+
+bool SourceCollection::SetSourceGain(SourceHandle source, float value)
+{
+    LOG_AUDIO->Debug("Source #{}: set gain {}", source, value);
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcef(static_cast<ALuint>(source), AL_GAIN, value);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set gain: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+std::array<float, 3> SourceCollection::GetSourcePosition(SourceHandle source) const
```

Comment:
Shouldn't it be static?

Reply:
Answered in https://github.com/Godlike/Tulpar/pull/6#discussion_r244268061